cql3/untyped_result_set.cc

```diff
@@ -47,58 +47,118 @@
 #include "result_set.hh"
 #include "transport/messages/result_message.hh"
 
-cql3::untyped_result_set_row::untyped_result_set_row(const map_t& data)
-    : _data(data)
+cql3::untyped_result_set_row::untyped_result_set_row(index_map_ptr index, const cql3::metadata& metadata, data_views data)
+    : _name_to_index(std::move(index))
+    , _metadata(metadata)
+    , _data(std::move(data))
 {}
 
-cql3::untyped_result_set_row::untyped_result_set_row(const std::vector<lw_shared_ptr<column_specification>>& columns, std::vector<bytes_opt> data)
-: _columns(columns)
-, _data([&columns, data = std::move(data)] () mutable {
-    map_t tmp;
-    std::transform(columns.begin(), columns.end(), data.begin(), std::inserter(tmp, tmp.end()), [](lw_shared_ptr<column_specification> c, bytes_opt& d) {
-       return std::make_pair<sstring, bytes_opt>(c->name->to_string(), std::move(d));
-    });
-    return tmp;
-}())
-{}
+size_t cql3::untyped_result_set_row::index(const std::string_view& name) const {
+    auto i = _name_to_index->find(name);
+    return i != _name_to_index->end() ? i->second : std::numeric_limits<size_t>::max();
+}
 
 bool cql3::untyped_result_set_row::has(std::string_view name) const {
-    auto i = _data.find(name);
-    return i != _data.end() && i->second;
+    auto i = index(name);
+    if (i < _data.size()) {
+        return !std::holds_alternative<std::monostate>(_data.at(i));
+    }
+    return false;
+}
+
+cql3::untyped_result_set_row::view_type cql3::untyped_result_set_row::get_view(std::string_view name) const {
+    return std::visit(make_visitor(
+        [](std::monostate) -> view_type { throw std::bad_variant_access(); },
+        [](const view_type& v) -> view_type { return v; },
+        [](const bytes& b) -> view_type { return view_type(b); }
+    ), _data.at(index(name)));
```

Comment:
I never know whether we should be using at() and pay for safety or [] and avoid checks. @avikivity what's your take on this?

Reply:
When things are not in hot path, I prefer the checked accessors, just to make sure we catch any borked code.

Reply:
Agree with @elcallio. These less-hot code paths are also more error-prone.

Reply:
Isn't this code executed at least once for each request? I thought that whatever gets executed 'per request' is already a hot path, no? I understand that we have a more hot path when something gets executed 'per sstable' but still once per request can be 1-5M times / s, no?

Reply:
That is not a hot path. An inner loop is a hot path.

Reply:
It seems that every person has its own definition of a hot path :). @tgrabiec was always telling me 'per request' operation is a hot path and I sort of agree considering that we're aiming for 5M ops/s

Reply:
That depends on the context. For a bounds check it is not. For a major allocation or prime number calculations it might be.